src/packet/bwe/arrival_group.rs

```diff
@@ -0,0 +1,309 @@
+use std::mem;
+use std::time::{Duration, Instant};
+
+use crate::rtp::SeqNo;
+
+use super::AckedPacket;
+
+const BURST_TIME_INTERVAL: Duration = Duration::from_millis(5);
+
+#[derive(Debug, Default)]
+pub struct ArrivalGroup {
+    first_seq_no: Option<SeqNo>,
+    last_seq_no: Option<SeqNo>,
+    first_local_send_time: Option<Instant>,
+    local_send_time: Option<Instant>,
+    remote_recv_time: Option<Instant>,
+    size: usize,
+}
+
+impl ArrivalGroup {
+    /// Maybe add a packet to the group.
+    ///
+    /// ## Return
+    ///
+    /// * [`Belongs::Belongs`] if the packet belongs to the group and was added.
+    /// * [`Belongs::NewGroup`] if the packet is the start of a new group and thus wasn't added.
+    ///                         The caller should create a new group and add this packet to the
+    ///                         group.
+    /// * [`Belongs::Skipped`] if the packet was skipped i.e. because it was out of order.
+    fn add_packet(&mut self, packet: AckedPacket) -> Belongs {
+        let belongs = self.belongs_to_group(&packet);
+
+        if belongs == Belongs::NewGroup || belongs == Belongs::Skipped {
+            return belongs;
+        }
+
+        self.remote_recv_time = Some(packet.remote_recv_time);
+
+        if self.first_local_send_time.is_none() {
+            self.first_local_send_time = Some(packet.local_send_time);
+            self.first_seq_no = Some(packet.seq_no);
+        }
+
+        self.local_send_time = self.local_send_time.max(Some(packet.local_send_time));
+        self.size += 1;
+        self.last_seq_no = Some(packet.seq_no);
+
+        Belongs::Belongs
+    }
+
+    fn belongs_to_group(&self, packet: &AckedPacket) -> Belongs {
+        let Some(first_local_send_time) = self.first_local_send_time else {
+            // Start of the group
+            return Belongs::Belongs;
+        };
+
+        let Some(send_diff) = packet
+            .local_send_time
+            .checked_duration_since(first_local_send_time) else {
+                warn!("Out of order send");
+                // Out of order
+                return Belongs::Skipped;
+        };
+
+        if send_diff < BURST_TIME_INTERVAL {
+            // Sent within the same burst interval
+            return Belongs::Belongs;
+        }
+
+        let inter_arrival_time = packet
+            .remote_recv_time
+            .checked_duration_since(self.remote_recv_time());
+
+        let Some(inter_arrival_time) = inter_arrival_time else {
+            warn!("Out of order arrival");
+            return Belongs::Skipped;
+        };
+
+        let inter_group_delay_variation = inter_arrival_time.as_secs_f64()
+            - (packet.local_send_time - self.local_send_time()).as_secs_f64();
```

Comment:
Might be more readable to have a make a "send_diff" variable for "packet.local_send_time - self.local_send_time()" and to use checked diff for it.

I'm not sure how to name it vs the other send_diff, though.

Reply:
I'll have a think